README.md
# MMM-Property-Managed

This is a module for the [MagicMirror²](https://github.com/MichMich/MagicMirror/).

Todo: Insert description here!

## Using the module

To use this module, add the following configuration block to the modules array in the `config/config.js` file:
```js
var config = {
    modules: [
        {
            module: 'MMM-Property-Managed',
            config: {
                // See below for configurable options
            }
        }
    ]
}
```

## Configuration options

| Option           | Description
|----------------- |-----------
| `token`        | *Required* Authentication token obtained from www.property-managed.com
| `useColor`        | *Optional* Specifies whether to use color in display. true (default) or false
| `updateInterval`        | *Optional* Interval to wait between updates (default 60000)
| `retryDelay`        | *Optional* Interval to wait between retries (default 5000)


CHANGELOG.md
# MMM-Property-Managed Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).


## [1.0.0]

First public release


node_helper.js
/* Magic Mirror
 * Node Helper: MMM-Property-Managed
 *
 * By Stewart McKee
 * MIT Licensed.
 */

var NodeHelper = require("node_helper");

module.exports = NodeHelper.create({

	// Override socketNotificationReceived method.

	/* socketNotificationReceived(notification, payload)
	 * This method is called when a socket notification arrives.
	 *
	 * argument notification string - The identifier of the noitication.
	 * argument payload mixed - The payload of the notification.
	 */
	socketNotificationReceived: function(notification, payload) {
		if (notification === "mmm-property-managed-update_event") {
			// Send notification
			this.sendNotificationTest(payload); //Is possible send objects :)
		}
	},

	// Example function send notification test
	sendNotificationTest: function(payload) {
		this.sendSocketNotification("mmm-property-managed-update_event", payload);
	}

});


propertyManagedMain.js
(function() {
  var template = Handlebars.template, templates = Handlebars.templates = Handlebars.templates || {};
templates['propertyManagedMain'] = template({"1":function(container,depth0,helpers,partials,data) {
    return "    <p class=\"bright\">Loading...</p>\n";
},"3":function(container,depth0,helpers,partials,data,blockParams,depths) {
    var stack1;

  return ((stack1 = helpers["if"].call(depth0 != null ? depth0 : (container.nullContext || {}),(depth0 != null ? depth0.noBuildingsWithActions : depth0),{"name":"if","hash":{},"fn":container.program(4, data, 0, blockParams, depths),"inverse":container.program(6, data, 0, blockParams, depths),"data":data})) != null ? stack1 : "");
},"4":function(container,depth0,helpers,partials,data) {
    return "      <p class=\"nothing-to-do bright\">Great! Nothing to do!</p>\n";
},"6":function(container,depth0,helpers,partials,data,blockParams,depths) {
    var stack1;

  return ((stack1 = helpers.each.call(depth0 != null ? depth0 : (container.nullContext || {}),(depth0 != null ? depth0.buildings : depth0),{"name":"each","hash":{},"fn":container.program(7, data, 0, blockParams, depths),"inverse":container.noop,"data":data})) != null ? stack1 : "");
},"7":function(container,depth0,helpers,partials,data,blockParams,depths) {
    var stack1;

  return ((stack1 = helpers["if"].call(depth0 != null ? depth0 : (container.nullContext || {}),(depth0 != null ? depth0.hasActions : depth0),{"name":"if","hash":{},"fn":container.program(8, data, 0, blockParams, depths),"inverse":container.noop,"data":data})) != null ? stack1 : "");
},"8":function(container,depth0,helpers,partials,data,blockParams,depths) {
    var stack1, helper, alias1=depth0 != null ? depth0 : (container.nullContext || {});

  return "          <div class=\"building\">\n            <header class=\"module_header bright\">"
    + container.escapeExpression(((helper = (helper = helpers.name || (depth0 != null ? depth0.name : depth0)) != null ? helper : helpers.helperMissing),(typeof helper === "function" ? helper.call(alias1,{"name":"name","hash":{},"data":data}) : helper)))
    + "</header>\n"
    + ((stack1 = helpers["if"].call(alias1,(depth0 != null ? depth0.urgentActions : depth0),{"name":"if","hash":{},"fn":container.program(9, data, 0, blockParams, depths),"inverse":container.noop,"data":data})) != null ? stack1 : "")
    + ((stack1 = helpers["if"].call(alias1,(depth0 != null ? depth0.upcomingActions : depth0),{"name":"if","hash":{},"fn":container.program(12, data, 0, blockParams, depths),"inverse":container.noop,"data":data})) != null ? stack1 : "")
    + "          </div>\n";
},"9":function(container,depth0,helpers,partials,data,blockParams,depths) {
    var stack1;

  return "              <div class=\"action-group\">\n                <h4 class=\""
    + container.escapeExpression(container.lambda((depths[1] != null ? depths[1].urgentClass : depths[1]), depth0))
    + "\">Urgent Actions</h4>\n                <table class=\"small\">\n"
    + ((stack1 = helpers.each.call(depth0 != null ? depth0 : (container.nullContext || {}),(depth0 != null ? depth0.urgentActions : depth0),{"name":"each","hash":{},"fn":container.program(10, data, 0, blockParams, depths),"inverse":container.noop,"data":data})) != null ? stack1 : "")
    + "                </table>\n              </div>\n";
},"10":function(container,depth0,helpers,partials,data) {
    var helper, alias1=depth0 != null ? depth0 : (container.nullContext || {}), alias2=helpers.helperMissing, alias3="function", alias4=container.escapeExpression;

  return "                    <tr class=\"normal\"><td class=\"action-name bright\">"
    + alias4(((helper = (helper = helpers.name || (depth0 != null ? depth0.name : depth0)) != null ? helper : alias2),(typeof helper === alias3 ? helper.call(alias1,{"name":"name","hash":{},"data":data}) : helper)))
    + "</td><td class=\"action-message\">"
    + alias4(((helper = (helper = helpers.message || (depth0 != null ? depth0.message : depth0)) != null ? helper : alias2),(typeof helper === alias3 ? helper.call(alias1,{"name":"message","hash":{},"data":data}) : helper)))
    + "</td></li>\n";
},"12":function(container,depth0,helpers,partials,data) {
    var stack1;

  return "              <div class=\"action-group\">\n                <h4 class=\"bright\">Upcoming Actions</h4>\n                <table class=\"small\">\n"
    + ((stack1 = helpers.each.call(depth0 != null ? depth0 : (container.nullContext || {}),(depth0 != null ? depth0.upcomingActions : depth0),{"name":"each","hash":{},"fn":container.program(10, data, 0),"inverse":container.noop,"data":data})) != null ? stack1 : "")
    + "                </table>\n              </div>\n";
},"compiler":[7,">= 4.0.0"],"main":function(container,depth0,helpers,partials,data,blockParams,depths) {
    var stack1;

  return "<div id=\"mmm-property-managed\">\n  <header class=\"module-header\">Property Managed</header>\n"
    + ((stack1 = helpers["if"].call(depth0 != null ? depth0 : (container.nullContext || {}),(depth0 != null ? depth0.loading : depth0),{"name":"if","hash":{},"fn":container.program(1, data, 0, blockParams, depths),"inverse":container.program(3, data, 0, blockParams, depths),"data":data})) != null ? stack1 : "")
    + "</div>\n";
},"useData":true,"useDepths":true});
})();

MMM-Property-Managed.js
/* global Module */

/* Magic Mirror
 * Module: MMM-Property-Managed
 *
 * By Stewart McKee
 * MIT Licensed.
 */


Module.register("MMM-Property-Managed", {
	defaults: {
		updateInterval: 60000,
		retryDelay: 5000,
    useColor: true
	},

	// requiresVersion: "2.1.0", // Required version of MagicMirror

	start: function() {
		var self = this;
		var dataRequest = null;
		var dataNotification = null;
		//Flag for check if module is loaded
		this.loaded = false;
		// Schedule update timer.
		this.getData();
		setInterval(function() {
			self.updateDom();
		}, this.config.updateInterval);
	},

	/*
	 * getData
	 * function example return data and show it in the module wrapper
	 * get a URL request
	 *
	 */
	getData: function() {
		var self = this;

		var urlApi = "https://www.property-managed.com/api/v1/buildings?token="+this.data.config.token;
		var retry = true;

		var dataRequest = new XMLHttpRequest();
		dataRequest.open("GET", urlApi, true);
		dataRequest.onreadystatechange = function() {
			if (this.readyState === 4) {
				if (this.status === 200) {
					self.processData(JSON.parse(this.response));
				} else if (this.status === 401) {
					self.updateDom(self.config.animationSpeed);
					Log.error(self.name, this.status);
					retry = false;
				} else {
					Log.error(self.name, "Could not load data.");
				}
				if (retry) {
					self.scheduleUpdate((self.loaded) ? -1 : self.config.retryDelay);
				}
			}
		};
		dataRequest.send();
	},


	/* scheduleUpdate()
	 * Schedule next update.
	 *
	 * argument delay number - Milliseconds before next update.
	 *  If empty, this.config.updateInterval is used.
	 */
	scheduleUpdate: function(delay) {
		var nextLoad = this.config.updateInterval;
		if (typeof delay !== "undefined" && delay >= 0) {
			nextLoad = delay;
		}
		nextLoad = nextLoad ;
		var self = this;
		setTimeout(function() {
			self.getData();
		}, nextLoad);
	},

  parseActions: function(actionsData) {
    var actions = [];

    if (actionsData.certificates) {
      for(var i in actionsData.certificates) {
        var certificate = actionsData.certificates[i];
        actions.push({"name": this.titleize(certificate.certificate_type), "message": "None Available"});
      }
    }

    if (actionsData.landlord_inspections_due) {
      actions.push({"name": "Landlord Inspection", "message": "Inspection Due"});
    }

    if (actionsData.alarms) {
      for(var j in actionsData.alarms) {
        var alarm = actionsData.alarms[j];
        actions.push({"name": alarm.location + " Alarm", "message": "Needs Replacing"});
      }
    }
    if (actionsData.invoices) {
      actions.push({"name": "Invoices", "message": "Outstanding"});
    }

    if (actionsData.rent_payments) {
      actions.push({"name": "Rent Payment", "message": "Overdue"});
    }
    if (actionsData.landlord_registration) {
      actions.push({"name": "Landlord Registration", "message": "None Available"});
    }
    return actions;
  },

  emptyActions: function(actions) {
    var empty = true;
    for(var key in actions) {
      if (actions[key] === undefined || actions[key] === null){
      }
      else {
        if (key !== "invoices" && typeof(actions[key]) == "boolean") {
          if (actions[key]) {
            empty = false;
          }
        } else if (typeof(actions[key]) == "object") {
          if (actions[key].length > 0) {
            empty = false;
          }
        }
      }


    }
    return empty;
  },

  titleize: function(str) {
    return str.replace("_"," ").split(" ").map(function(word) {    
      return word.substr(0, 1).toUpperCase() + word.substr(1, word.length);
    }).join(" ")
  },
  getDom: function() {

    var buildings = [];

    var loading = (this.dataNotification === undefined)
    var noBuildingsWithActions = true;

    for(var i in this.dataNotification) {
      var buildingData = this.dataNotification[i];
      delete buildingData.urgent_actions.invoices;
      delete buildingData.upcoming_actions.invoices;
      var data = {};
      data.name = buildingData.name;
      if (!this.emptyActions(buildingData.urgent_actions))
        data.urgentActions = this.parseActions(buildingData.urgent_actions);
      if (!this.emptyActions(buildingData.upcoming_actions))
        data.upcomingActions = this.parseActions(buildingData.upcoming_actions);

      if (data.urgentActions || data.upcomingActions) {
        data.hasActions = true;
        noBuildingsWithActions = false;
      } else {
        data.hasActions = false;
      }

      buildings.push(data);
    }


    var context = {
      buildings: buildings,
      noBuildingsWithActions: noBuildingsWithActions,
      loading: loading
    };

    if (this.config.useColor) {
      context.urgentClass = "red";
    } else {
      context.urgentClass = "bright";
    }

    var template = Handlebars.templates.propertyManagedMain;
    var parser = new DOMParser();
    return parser.parseFromString(template(context), "text/html").getElementById("mmm-property-managed");
  },
	getScripts: function() {
		return [
      "https://cdnjs.cloudflare.com/ajax/libs/handlebars.js/4.0.10/handlebars.min.js",
      "propertyManagedMain.js"
    ];
	},

	getStyles: function () {
		return [
			"MMM-Property-Managed.css",
		];
	},

	// Load translations files
	getTranslations: function() {
		//FIXME: This can be load a one file javascript definition
		return {
			en: "translations/en.json",
			es: "translations/es.json"
		};
	},

	processData: function(data) {
		var self = this;
		this.dataRequest = data;
		if (this.loaded === false) { self.updateDom(self.config.animationSpeed) ; }
		this.loaded = true;
		// the data if load
		// send notification to helper
		this.sendSocketNotification("mmm-property-managed-update_event", data);
	},

	// socketNotificationReceived from helper
	socketNotificationReceived: function (notification, payload) {
		if(notification === "mmm-property-managed-update_event") {
			// set dataNotification
			this.dataNotification = payload;
			this.updateDom();
		}
	},
});
